js/modules/types/contact.js

```diff
@@ -0,0 +1,142 @@
+const { omit, compact, map } = require('lodash');
+
+const { toLogFormat } = require('./errors');
+const { SignalService } = require('../../../ts/protobuf');
+
+const DEFAULT_PHONE_TYPE = SignalService.DataMessage.Contact.Phone.Type.HOME;
+const DEFAULT_EMAIL_TYPE = SignalService.DataMessage.Contact.Email.Type.HOME;
+const DEFAULT_ADDRESS_TYPE =
+  SignalService.DataMessage.Contact.PostalAddress.Type.HOME;
+
+exports.parseAndWriteContactAvatar = upgradeAttachment => async (
+  contact,
+  context = {}
+) => {
+  const { message } = context;
+  const { avatar } = contact;
+  const contactWithUpdatedAvatar =
+    avatar && avatar.avatar
+      ? Object.assign({}, contact, {
+          avatar: Object.assign({}, avatar, {
+            avatar: await upgradeAttachment(avatar.avatar, context),
+          }),
+        })
+      : omit(contact, ['avatar']);
+
+  // eliminates empty numbers, emails, and addresses; adds type if not provided
+  const contactWithCleanedElements = parseContact(contactWithUpdatedAvatar);
+
+  const error = exports._validateContact(contactWithCleanedElements, {
+    messageId: idForLogging(message),
+  });
+  if (error) {
+    console.log(
+      'Contact.parseAndWriteContactAvatar: contact was malformed.',
+      toLogFormat(error)
+    );
+  }
+
+  return contactWithCleanedElements;
+};
+
+function parseContact(contact) {
+  return Object.assign(
+    {},
+    omit(contact, ['avatar', 'number', 'email', 'address']),
+    parseAvatar(contact.avatar),
+    createArrayKey('number', compact(map(contact.number, cleanPhoneItem))),
+    createArrayKey('email', compact(map(contact.email, cleanEmailItem))),
+    createArrayKey('address', compact(map(contact.address, cleanAddress)))
+  );
+}
+
+function idForLogging(message) {
+  return `${message.source}.${message.sourceDevice} ${message.sent_at}`;
+}
+
+exports._validateContact = (contact, options = {}) => {
```

Comment:
Ditto: Now that these are in their own module, we can probably omit the `*Contact` in the API names 👀